src/devices/SenseHat/SenseHatLedMatrix.cs

```diff
@@ -95,9 +144,300 @@ public static int PositionToIndex(int x, int y)
+            {
+                _textAnimationTimer.Stop();
+                _textAnimationTimer.Elapsed -= TextAnimationTimer_Elapsed;
+                _textAnimationTimer.Dispose();
+                _textAnimationTimer = null;
+            }
+        }
+
+        private void TextAnimationTimer_Elapsed(object? sender, System.Timers.ElapsedEventArgs e)
+        {
+            SenseHatTextRenderState? renderState;
+            // Lock because render state is nullable and assignment may not be atomic
+            lock (_lockTextRenderState)
+            {
+                renderState = _textRenderState;
+            }
+
+            if (renderState != null)
+            {
+                var renderMatrix = renderState.TextRenderMatrix;
+                if (renderMatrix.Text.Length > 1)
+                {
+                    renderMatrix.ScrollByOnePixel();
+                    RenderText(renderState);
+                }
+            }
+        }
+
+        private void RenderText(SenseHatTextRenderState? textRenderState)
+        {
+            if (textRenderState == null)
+            {
+                return;
+            }
+
+            // Allocate frame buffer to hold color values for one frame
+            var frameBuffer = new Color[NumberOfPixelsPerColumn * NumberOfPixelsPerRow];
+
+            // Render the 8x8 matrix
+            for (var x = 0; x < NumberOfPixelsPerColumn; x++)
+            {
+                for (var y = 0; y < NumberOfPixelsPerRow; y++)
+                {
+                    int tx = x;
+                    int ty = y;
+                    switch (textRenderState.TextRotation)
+                    {
+                        case SenseHatTextRotation.Rotate_0_Degrees:
+                            break;
+                        case SenseHatTextRotation.Rotate_90_Degrees:
+                            tx = y;
+                            ty = NumberOfPixelsPerColumn - x - 1;
+                            break;
+                        case SenseHatTextRotation.Rotate_180_Degrees:
+                            tx = NumberOfPixelsPerColumn - x - 1;
+                            ty = NumberOfPixelsPerRow - y - 1;
+                            break;
+                        case SenseHatTextRotation.Rotate_270_Degrees:
+                            tx = NumberOfPixelsPerColumn - y - 1;
+                            ty = x;
+                            break;
+                    }
+
+                    var frameIndex = PositionToIndex(tx, ty);
+
+                    if (textRenderState.TextRenderMatrix.IsPixelSet(x, y))
+                    {
+                        frameBuffer[frameIndex] = textRenderState.TextColor;
+                    }
+                    else
+                    {
+                        frameBuffer[frameIndex] = textRenderState.TextBackgroundColor;
+                    }
+                }
+            }
+
+            // Lock write so that SenseHat disposal does not interfere
```

Comment:
it's ok if someone disposes matrix prematurely as long as you get reasonable exception - we do not want to pollute the code just because someone might do weird things with the hardware

Reply:
It's the disposal of the I2C bus that interferes in this case. Currently this works without exception:

```
using (var sh = new SenseHat()) 
{
    sh.LedMatrix.SetText("abc");
    // ... do something
}
```

In that case the text scrolling automatically starts because the text is long. At the end of the `using`, the `SenseHat` disposes the I2C bus. Without precautions, the text rendering could attempt to access the disposed I2C bus because it is in the middle of rendering a new frame. The `Lock` around `Write` in `SenseHatLedMatrix` prevents access to a disposed I2C device.

With the current API design (text automatically scrolls) it would not be obvious to the user that something needs to be reset before disposing the `SenseHat`. One could give up simplicity and ask the user to explicitly enable and disable text scrolling. Or demand that text is set to 'empty' before disposing. Or any exception in the text rendering method could be ignored. My preference would be to retain the lock for the simplest use without surprises; this would be for SenseHat on Pi with its small LED matrix. It is possible that generalizations for other LED displays require a different rendering method or make other design choices, eliminating the lock in their approach. - What do you think?